pubsub/redis/redis.go

```diff
@@ -113,68 +180,215 @@ func (r *redisStreams) Publish(req *pubsub.PublishRequest) error {
 
 func (r *redisStreams) Subscribe(req pubsub.SubscribeRequest, handler func(msg *pubsub.NewMessage) error) error {
 	err := r.client.XGroupCreateMkStream(req.Topic, r.metadata.consumerID, "0").Err()
-	if err != nil {
-		r.logger.Warnf("redis streams: %s", err)
+	// Ignore BUSYGROUP errors
+	if err != nil && err.Error() != "BUSYGROUP Consumer Group name already exists" {
+		r.logger.Errorf("redis streams: %s", err)
+
+		return err
 	}
-	go r.beginReadingFromStream(req.Topic, r.metadata.consumerID, handler)
+
+	go r.pollMessagesLoop(req.Topic, handler)
+	go r.reclaimMessagesLoop(req.Topic, handler)
 
 	return nil
 }
 
-func (r *redisStreams) readFromStream(stream, consumerID, start string) ([]redis.XStream, error) {
-	res, err := r.client.XReadGroup(&redis.XReadGroupArgs{
-		Group:    consumerID,
-		Consumer: consumerID,
-		Streams:  []string{stream, start},
-		Block:    0,
-	}).Result()
-	if err != nil {
-		return nil, err
-	}
+func (r *redisStreams) enqueue(stream, consumerID string, handler func(msg *pubsub.NewMessage) error, msgs []redis.XMessage) {
```

Comment:
Please, call out this refactoring as a separate issue with release notes.

Reply:
Created https://github.com/dapr/components-contrib/issues/710